---
title: "Sweat electrolytes and CFTR activity in the era of CFTR modulators"
author: "Tim Vigers"
date: "today"
date-format: long
format:
  html:
    toc: true
    toc-depth: 10
    toc-float: true
    number-sections: true
    code-fold: true
    self-contained: true
    fig-cap-location: top
    page-layout: full
    theme:
      light: flatly
      dark: darkly
editor: source
---

```{r setup}
#| include: false
library(tidyverse)
library(readxl)
library(gtsummary)
home_dir <- switch(Sys.info()["sysname"],
  "Darwin" = "/Users/timvigers/Library/CloudStorage/OneDrive-TheUniversityofColoradoDenver/Vigers/CF/Edith Zemanick/Sweat Test",
  "Windows" = "C:/Users/timvigers/OneDrive - The University of Colorado Denver/Vigers/CF/Edith Zemanick/Sweat Test",
  "Linux" = "/home/timvigers/OneDrive/Vigers/CF/Edith Zemanick/Sweat Test"
)
knitr::opts_knit$set(root.dir = home_dir)
```

```{r data cleaning}
#| message: false
# Import data
chco <- read.csv("./Data_Raw/Copy of Sweat Tests_AllFor202558_revised01Feb2021.csv",
  na.strings = ""
)
# Numeric variables
num_vars <- c(
  "Na", "K", "Cl", "Amount_Left", "Amount_Right", "Na_Left", "Na_Right",
  "K_Left", "K_Right", "Cl_Left", "Cl_Right"
)
chco[, num_vars] <- lapply(chco[, num_vars], function(c) {
  suppressWarnings(as.numeric(c))
})
# Fix diagnoses/exclusion data for CHCO (based on Angela's review)
diagnoses <- read_excel("./Data_Cleaned/AM reclassifications All Patients with CF or Non-CF.xlsx")
exclude <- diagnoses$PatientID[!is.na(diagnoses$Exclude) |
  diagnoses$Code == "Exclude"]
exclude <- exclude[!is.na(exclude)]
chco <- chco[!(chco$PatientID %in% exclude), ]
chco$Diagnosis <- factor(
  diagnoses$Diagnosis[match(
    chco$PatientID,
    diagnoses$PatientID
  )],
  levels = c("CFTR-related disorder", "CRMS", "Cystic Fibrosis", "Non-CF"),
  labels = c("CFTR-RD or CRMS", "CFTR-RD or CRMS", "Cystic Fibrosis", "Non-CF")
)
# Exclude those without demographics (not confident in diagnosis per Edith)
chco <- chco %>% filter(Demographics.Available == "TRUE")
# Limit to the columns that we need for analysis
chco <- chco %>% select(
  PatientID, Site, Diagnosis, Sex, Genotypes1, Genotypes2,
  Age.at.Test.in.Years, Method, Na, K, Cl, Na_Left, Na_Right, K_Left,
  K_Right, Cl_Left, Cl_Right
)
# Import NJH data
njh <- read_excel("./Data_Raw/SweatVariability2025_CFgenotypes 2024-08-21.xlsx")
# Filter and select similar to CHCO
njh <- njh %>%
  filter(`Demographics Available`) %>%
  mutate(Site = "NJH") %>%
  select(
    `Patient ID`, `Site of test`, Diagnosis, Sex, `Genotype 1`, `Genotype 2`,
    `Age at test (in years)`, `Test Method`, `Sweat sodium`,
    `Sweat potassium`, `Sweat chloride`, `Sweat sodium left arm`,
    `Sweat sodium right arm`, `Sweat potassium left arm`,
    `Sweat potassium right arm`, `Sweat chloride left arm`,
    `Sweat chloride right arm`
  )
colnames(njh) <- colnames(chco)
njh <- njh %>%
  group_by(PatientID) %>%
  slice_min(Age.at.Test.in.Years, na_rm = T)
# Import detailed diagnoses from Katie
diagnoses <- read.csv("./Data_Cleaned/SweatVariability2025_CFgenotypes.csv",
  na.strings = c("", ".")
)
# Merge NJH and diagnoses
diagnoses <- diagnoses %>% select(Patient.ID, Diagnosis...DataScout)
colnames(diagnoses) <- c("PatientID", "Diagnosis")
# Less detailed diagnoses
diagnoses$Diagnosis <- factor(diagnoses$Diagnosis,
  levels = c(
    "Abnormal sweat choride test without diagnosis of cystic fibrosis",
    "Abnormal sweat test",
    "Alpha-1-antitrypsin deficiency",
    "Alpha-1-antitrypsin deficiency carrier",
    "Alpha-1-antitrypsin deficiency, Positive cystic fibrosis sweat test without diagnosis of cystic fibrosis",
    "CF", "CF carrier, Alpha-1-antitrypsin deficiency",
    "CF, alpha 1-antitrypsin deficiency",
    "CF, alpha 1-antitrypsin deficiency carrier",
    "CF, Alpha-1-antitrypsin deficiency",
    "Cystic fibrosis carrier",
    "Cystic fibrosis carrier, Alpha-1-antitrypsin deficiency",
    "Cystic fibrosis carrier, Alpha-1-antitrypsin deficiency carrier",
    "Cystic fibrosis screening",
    "Cystic fibrosis transmembrane conductance regulator (CFTR)-related disorder",
    "Positive cystic fibrosis sweat test without diagnosis of cystic fibrosis",
    "Primary ciliary dyskinesia",
    "Primary Ciliary dyskinesia, Cystic Fibrosis carrier"
  ),
  labels = c(
    "Non-CF", "Non-CF", "Non-CF", "CF Carrier", "Non-CF", "Cystic Fibrosis",
    "CF Carrier", "Cystic Fibrosis", "Cystic Fibrosis", "Cystic Fibrosis",
    "CF Carrier", "CF Carrier", "CF Carrier", "Non-CF", "CFTR-RD or CRMS",
    "Non-CF", "Non-CF", "CF Carrier"
  )
)
njh <- full_join(njh, diagnoses)
# Import CHEC data
chec <- read.csv("./Data_Raw/CHEC sweat sodium and chloride.csv")
# Filter and select as above (need to add some empty columns and split genotype)
chec$Genotypes1 <- trimws(sapply(strsplit(chec$gntyp_pair, "/"), "[[", 1))
chec$Genotypes2 <- trimws(sapply(strsplit(chec$gntyp_pair, "/"), "[[", 2))
chec$Diagnosis <- "Cystic Fibrosis"
chec[, c(
  "Method", "K", "Na_Left", "Na_Right", "K_Left", "K_Right", "Cl_Left",
  "Cl_Right"
)] <- NA
chec <- chec %>%
  select(
    subject, site, Diagnosis, sex, Genotypes1, Genotypes2, age, Method,
    swsavg, K, swcavg, Na_Left, Na_Right, K_Left, K_Right, Cl_Left, Cl_Right
  )
colnames(chec) <- colnames(chco)
# Put everything together
df <- full_join(chco, njh)
df$PatientID <- as.character(df$PatientID)
df <- full_join(df, chec)
# Average the left and right values
df$NaAvg <- rowMeans(df[, c("Na_Left", "Na_Right")], na.rm = T)
df$NaAvg[is.nan(df$NaAvg)] <- NA
df$KAvg <- rowMeans(df[, c("K_Left", "K_Right")], na.rm = T)
df$KAvg[is.nan(df$KAvg)] <- NA
df$ClAvg <- rowMeans(df[, c("Cl_Left", "Cl_Right")], na.rm = T)
df$ClAvg[is.nan(df$ClAvg)] <- NA
df <- df %>%
  unite(Na, c("Na", "NaAvg"), na.rm = T) %>%
  unite(K, c("K", "KAvg"), na.rm = T) %>%
  unite(Cl, c("Cl", "ClAvg"), na.rm = T)
# Genotype function
genotype_function <- read.csv("./Data_Cleaned/genotype_function.csv")
# Get single genes and add them together
genotype_function <-
  genotype_function[grep("NULL", genotype_function$genotype), ]
genotype_function$genotype <- sub("/NULL", "", genotype_function$genotype)
df$Genotype1Function <-
  genotype_function$percent_wildtype_cftr_function[match(
    df$Genotypes1,
    genotype_function$genotype
  )]
df$Genotype2Function <-
  genotype_function$percent_wildtype_cftr_function[match(
    df$Genotypes2,
    genotype_function$genotype
  )]
df$TotalGenotypeFunction <- rowSums(
  df[, c("Genotype1Function", "Genotype2Function")]
)
df$TotalGenotypeFunction[df$TotalGenotypeFunction == 0] <- NA
# Numeric variables
df[, c("Na", "K", "Cl")] <- lapply(df[, c("Na", "K", "Cl")], as.numeric)
# As data frame
df <- as.data.frame(df)
# Get rid of unnecessary columns
df <- df %>%
  select(
    PatientID, Site, Diagnosis, Sex, Age.at.Test.in.Years, Na:Cl,
    Genotypes1, Genotype1Function, Genotypes2, Genotype2Function,
    TotalGenotypeFunction
  )
# For each ID, get the row with the most data (or first test)
df$Missing <- rowSums(is.na(df))
df <- df %>%
  group_by(PatientID) %>%
  slice_min(Missing) %>%
  slice_min(row_number()) %>%
  filter(!is.na(Diagnosis)) %>%
  ungroup()
# Set non-CF to 100% CFTR function (200 because 100 for each allele)
df$TotalGenotypeFunction[df$Diagnosis == "Non-CF"] <- 200
# Write for checking
write.csv(df, "./Data_Cleaned/final_analysis_dataset.csv",
  row.names = F, na = ""
)
```

```{r}
#| message: false
#| warning: false
df %>%
  ggplot(
    aes(x = Na, y = TotalGenotypeFunction, color = Diagnosis, group = Diagnosis)
  ) +
  geom_point() +
  theme()
df %>%
  ggplot(
    aes(x = Cl, y = TotalGenotypeFunction, color = Diagnosis, group = Diagnosis)
  ) +
  geom_point() +
  theme()
df %>%
  ggplot(
    aes(x = Na / Cl, y = TotalGenotypeFunction, color = Diagnosis, group = Diagnosis)
  ) +
  geom_point() +
  theme()
```

```{r}
df %>%
  filter(Diagnosis == "Cystic Fibrosis") %>%
  mutate(
    `Age group` = cut(Age.at.Test.in.Years, c(-Inf, 18, Inf), right = F),
    `Cl < 60` = Cl < 60
  ) %>%
  select(Na, Cl, TotalGenotypeFunction, `Cl < 60`, `Age group`) %>%
  tbl_summary(by = `Age group`, missing_text = "Missing") %>%
  add_p()
```
